driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,62 @@ public final class CommandMessage extends RequestMessage {
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> byteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf byteBuf = new CompositeByteBuf(byteBuffers);
+            try {
+                byteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+                ByteBufBsonDocument byteBufBsonDocument = createOne(byteBuf);
+
+                // If true, it means there is at least one Kind 1:Document Sequence in the OP_MSG
+                if (byteBuf.hasRemaining()) {
+                    BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
+
+                    // Each loop iteration processes one Document Sequence
+                    // When there are no more bytes remaining, there are no more Document Sequences
+                    while (byteBuf.hasRemaining()) {
+                        byteBuf.position(byteBuf.position() + 1 /* payload type */ + 4 /* payload size */);
+                        String payloadName = getPayloadName(byteBuf);
+                        assertFalse(payloadName.contains("."));
+                        commandBsonDocument.append(payloadName, new BsonArray(createList(byteBuf)));
+                    }
+                    return commandBsonDocument;
+                } else {
+                    return byteBufBsonDocument;
+                }
+            } finally {
+                byteBuf.release();
+            }
+        } finally {
+            byteBuffers.forEach(ByteBuf::release);
         }
-
-       return commandBsonDocument;
     }
 
-    boolean containsPayload() {
-        return payload != null;
+    /**
+     * Get the field name from a buffer positioned at the start of the document sequence identifier of an OP_MSG Section of type
+     * Document Sequence (Kind 1).
+     * <p>
+     * Upon normal completion of the method, the buffer will be positioned at the start of the first BSON object in the sequence.
+    */
+    private String getPayloadName(final ByteBuf outputByteBuf) {
+        List<Byte> payloadNameBytes = new ArrayList<>();
+        byte curByte = outputByteBuf.get();
+        while (curByte != 0) {
+            payloadNameBytes.add(curByte);
+            curByte = outputByteBuf.get();
+        }
+
+        // Convert List<Byte> to byte[]
+        byte[] byteArray = new byte[payloadNameBytes.size()];
+        IntStream.range(0, payloadNameBytes.size()).forEachOrdered(i -> byteArray[i] = payloadNameBytes.get(i));
+        return new String(byteArray, StandardCharsets.UTF_8);
     }
```

Comment:
I agree with @NathanQingyangXu's [suggestion](https://github.com/mongodb/mongo-java-driver/pull/1456/files?show-viewed-files=true&file-filters%5B%5D=#r1688958770). The code will look as follows, I believe (unfortunately, we have to deal with `UnsupportedEncodingException` because the `toString` method that does not throw it exists only since Java SE 10):

```suggestion
    private String getSequenceIdentifier(final ByteBuf byteBuf) {
        ByteArrayOutputStream sequenceIdentifierBytes = new ByteArrayOutputStream();
        byte curByte = byteBuf.get();
        while (curByte != 0) {
            sequenceIdentifierBytes.write(curByte);
            curByte = byteBuf.get();
        }
        try {
            return sequenceIdentifierBytes.toString(StandardCharsets.UTF_8.name());
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }
```

Reply:
I committed and pushed this change locally, as I wanted to change one small thing about the suggestion: generally I think we throw MongoInternalException in these situations instead of RuntimeException.